src/Databases/PostgreSQL/DatabasePostgreSQLReplica.cpp

```diff
@@ -0,0 +1,234 @@
+        const String & postgres_database_name,
+        PostgreSQLConnectionPtr connection_,
+        std::unique_ptr<PostgreSQLReplicaSettings> settings_)
+    : DatabaseOrdinary(
+            database_name_, metadata_path_, "data/" + escapeForFileName(database_name_) + "/",
+            "DatabasePostgreSQLReplica<Ordinary> (" + database_name_ + ")", context)
+    , log(&Poco::Logger::get("PostgreSQLReplicaDatabaseEngine"))
+    , global_context(context.getGlobalContext())
+    , metadata_path(metadata_path_)
+    , database_engine_define(database_engine_define_->clone())
+    , database_name(database_name_)
+    , remote_database_name(postgres_database_name)
+    , connection(std::move(connection_))
+    , settings(std::move(settings_))
+{
+}
+
+
+template<>
+DatabasePostgreSQLReplica<DatabaseAtomic>::DatabasePostgreSQLReplica(
+        const Context & context,
+        const String & metadata_path_,
+        UUID uuid,
+        const ASTStorage * database_engine_define_,
+        const String & database_name_,
+        const String & postgres_database_name,
+        PostgreSQLConnectionPtr connection_,
+        std::unique_ptr<PostgreSQLReplicaSettings> settings_)
+    : DatabaseAtomic(database_name_, metadata_path_, uuid, "DatabasePostgreSQLReplica<Atomic> (" + database_name_ + ")", context)
+    , global_context(context.getGlobalContext())
+    , metadata_path(metadata_path_)
+    , database_engine_define(database_engine_define_->clone())
+    , remote_database_name(postgres_database_name)
+    , connection(std::move(connection_))
+    , settings(std::move(settings_))
+{
+}
+
+
+template<typename Base>
+void DatabasePostgreSQLReplica<Base>::startSynchronization()
+{
+    replication_handler = std::make_unique<PostgreSQLReplicationHandler>(
+            remote_database_name,
+            connection->conn_str(),
+            metadata_path + METADATA_SUFFIX,
+            std::make_shared<Context>(global_context),
+            settings->postgresql_replica_max_block_size.changed
+                     ? settings->postgresql_replica_max_block_size.value
+                     : (global_context.getSettingsRef().max_insert_block_size.value),
+            settings->postgresql_replica_allow_minimal_ddl, true,
+            settings->postgresql_replica_tables_list.value);
+
+    std::unordered_set<std::string> tables_to_replicate = replication_handler->fetchRequiredTables(connection->conn());
+
+    for (const auto & table_name : tables_to_replicate)
+    {
+        auto storage = getStorage(table_name);
+
+        if (storage)
+        {
+            replication_handler->addStorage(table_name, storage->template as<StoragePostgreSQLReplica>());
+            tables[table_name] = storage;
+        }
+    }
+
+    LOG_TRACE(log, "Loaded {} tables. Starting synchronization", tables.size());
+    replication_handler->startup();
+}
+
+
+template<typename Base>
+StoragePtr DatabasePostgreSQLReplica<Base>::getStorage(const String & name)
```

Comment:
> Seems like the only place where we use it is DatabasePostgreSQLReplica<Base>::startSynchronization()

Fixed. 

> What will happen if tryGetTable(...) returned some table (not nullptr, so table exists) and we try to add existing table again in startSynchronization()?

You mean whether it is possible that synchronization thread in replication handler might attempt to add tables if they already exist? It will not add new tables if replication slot was created and tables were loaded, there is a check for that. But I need to add a test.